main.go

```diff
@@ -77,188 +74,107 @@ func extractFromTag(tag *ast.BasicLit) (string, string) {
 }
 
 // Get v1beta1.JSONSchemaProps given ast.Expr
-func newDefinition(t ast.Expr, comment string, importPaths map[string]string, curPkgPrefix string) (*v1beta1.JSONSchemaProps, []TypeReference) {
+func exprToSchema(t ast.Expr, comment string, importPaths map[string]string, curPkgPrefix string) (*v1beta1.JSONSchemaProps, []TypeReference) {
 	def := &v1beta1.JSONSchemaProps{}
 	externalTypeRefs := []TypeReference{}
+	var typeName, pkgAlias string
 
 	def.Description = comment
 
 	switch tt := t.(type) {
 	case *ast.Ident:
-		typeName := tt.Name
-		if isSimpleType(typeName) {
-			def.Type = jsonifyType(typeName)
-		} else {
-			def.Ref = getPrefixedDefLink(typeName, curPkgPrefix)
-		}
+		typeName = tt.Name
 	case *ast.ArrayType:
-		arrayType := tt
-		var typeNameOfArray, packageAlias string
-		switch arrayType.Elt.(type) {
-		case *ast.Ident:
-			identifier := arrayType.Elt.(*ast.Ident)
-			typeNameOfArray = identifier.Name
-		case *ast.StarExpr:
-			starType := arrayType.Elt.(*ast.StarExpr)
-			identifier := starType.X.(*ast.Ident)
-			typeNameOfArray = identifier.Name
-		case *ast.SelectorExpr:
-			selectorType := arrayType.Elt.(*ast.SelectorExpr)
-			packageAlias = selectorType.X.(*ast.Ident).Name
-			typeName := selectorType.Sel.Name
-			typeNameOfArray = typeName
-			externalTypeRefs = append(externalTypeRefs, TypeReference{typeName, packageAlias})
-		}
+		props, extTypeRefs := arrayTypeToSchema(tt, comment, importPaths, curPkgPrefix)
+		return props, extTypeRefs
+	case *ast.MapType:
+		mapTypeToSchema(tt, def, curPkgPrefix)
+	case *ast.SelectorExpr:
+		selectorType := tt
+		pkgAlias = selectorType.X.(*ast.Ident).Name
+		typeName = selectorType.Sel.Name
 
-		if def.Items == nil {
-			def.Items = &v1beta1.JSONSchemaPropsOrArray{}
-		}
-		def.Items.Schema = new(v1beta1.JSONSchemaProps)
-		if isSimpleType(typeNameOfArray) {
-			def.Items.Schema.Type = jsonifyType(typeNameOfArray)
+		def.Ref = getPrefixedDefLink(typeName, importPaths[pkgAlias])
+		externalTypeRefs = []TypeReference{{TypeName: typeName, PackageName: pkgAlias}}
+	case *ast.StarExpr:
+		def, externalTypeRefs = exprToSchema(tt.X, comment, importPaths, curPkgPrefix)
+	case *ast.StructType:
+		externalTypeRefs, def = structTypeToSchema(tt, importPaths, curPkgPrefix, def)
```

Comment:
For consistency, (at least for now) can we stick to `def, externalTypeRefs` as written above?